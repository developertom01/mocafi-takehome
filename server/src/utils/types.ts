
// Map a particular string to a type
export type LocalStorage = {
  requestId: string;
};
