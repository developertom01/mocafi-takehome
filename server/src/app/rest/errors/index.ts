import { BaseError } from "../../../errors/base-error";

export class HttpBaseError extends BaseError {
  constructor(
    public readonly statusCode: number,
    message: string,
    data?: object | string
  ) {
    super(message, data);
  }
}
