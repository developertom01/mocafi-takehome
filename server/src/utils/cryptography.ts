import fs from "fs/promises";
import { createHmac, timingSafeEqual } from "crypto";
import { APP_ENV, MASTER_KEY_PATH } from "../config/app-config";

function validateKey(key: string | Buffer, size: number = 32) {
  if (!Buffer.isBuffer(key)) {
    key = Buffer.from(key, "hex");
  }
  // Check if key length has sufficient bits
  if (key.length !== size) {
    throw new Error(`Invalid key length: Key must be ${size} bytes long`);
  }
}

export function hash(data: string, secretKey: string) {
  validateKey(secretKey);
  return createHmac("sha256", secretKey).update(data).digest("hex");
}

export function verifyHash(plainText: string, hash: string, secretKey: string) {
  const newHash = createHmac("sha256", secretKey).update(plainText).digest();

  return timingSafeEqual(newHash, Buffer.from(hash));
}

export async function getEncryptionMastKey(filePath: string) {
  if (APP_ENV === "production") return;

  try {
    await fs.access(filePath);
  } catch (error) {
    throw new Error("Encryption master key not found");
  }

  const key = await fs.readFile(filePath, "base64");

  return key;
}

export async function getLocalKmsProvider() {
  const masterKey = await getEncryptionMastKey(MASTER_KEY_PATH);
  return {
    local: {
      key: masterKey!,
    },
  };
}

function getAzureKmsProvider() {
  return {
    azure: {
      tenantId: "myTenant",
      clientId: "myClient",
      clientSecret: "mySecret",
    },
  };
}

export async function getKmsProvider() {
  return APP_ENV === "production"
    ? getAzureKmsProvider()
    : await getLocalKmsProvider();
}
