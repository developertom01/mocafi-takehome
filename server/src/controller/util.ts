import { BaseError } from "../errors/base-error";
import { MongoServerError } from "mongodb";
import { ValidationError } from "../errors/validation-error";
import { UncaughtError } from "../errors/uncaught-error";

export function handleDataLayerError(err: Error) {
  if (err instanceof BaseError) {
    return err;
  }
  // Handle mongodb errors
  if (err instanceof MongoServerError) {
    switch (err.code) {
      case 11000:
        return new ValidationError(
          `${Object.keys(err.keyValue)[0]} already exists`,
          {
            ...err.keyValue,
          }
        );
      default:
        return new UncaughtError("Database error", err);
    }
  }

  return new UncaughtError("Uncaught error", err);
}
