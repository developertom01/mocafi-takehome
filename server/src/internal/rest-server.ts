import express, { Application, NextFunction } from "express";
import swaggerUi from "swagger-ui-express";
import swaggerConfig from "../config/swagger-config";

import cors from "cors";
import { Logger } from "./logger";
import { AsyncLocalStorage } from "async_hooks";
import { LocalStorage } from "../utils/types";

export interface RestServer<T> {
  app: T;
}

export class ExpressRestServer implements RestServer<Application> {
  private readonly _app: Application;

  constructor(
    apiRouter: express.Router,
    localStorage: AsyncLocalStorage<LocalStorage>,
    private readonly logger: Logger
  ) {
    this._app = express();
    this._app.use(express.json());
    this._app.use(cors());

    // Initialize local storage and set request id
    this._app.use((_, __, next) => {
      if (!localStorage.getStore()) {
        localStorage.run(
          {
            requestId: crypto.randomUUID(),
          },
          () => {
            next();
          }
        );
      } else {
        next();
      }
    });

    this._app.all("/", (_, res) => {
      res.send("Hello World!");
    });

    this._app.use("/api", apiRouter);
    this._app.use("/api-doc", swaggerUi.serve, swaggerUi.setup(swaggerConfig));
    this._app.use((err: Error, _: any, res: any, __: NextFunction) => {
      this.logger.error(err);
      res.status(500).send("Something broke!");
    });
  }

  public get app() {
    return this._app;
  }
}
