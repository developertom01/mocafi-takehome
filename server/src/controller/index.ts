import { MongoClient } from "mongodb";
import { Logger } from "../internal/logger";
import {
  UserAccountController,
  UserAccountControllerType,
} from "./user-accounts";

export interface Controller {
  userAccountController: UserAccountControllerType;
}

export function setupController(db: MongoClient, logger: Logger): Controller {
  return {
    userAccountController: new UserAccountController(db, logger),
  };
}
