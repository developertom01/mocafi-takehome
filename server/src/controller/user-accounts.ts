import zod from "zod";

import { MongoClient, ObjectId } from "mongodb";
import { Logger } from "../internal/logger";
import { APP_SECRET, DATABASE_NAME } from "../config/app-config";
import { handleDataLayerError } from "./util";
import { ValidationError } from "../errors/validation-error";
import { NotFoundError } from "../errors/not-found";
import {
  decryptField,
  encryptField,
  hash,
  verifyHash,
} from "../utils/cryptography";
import { UnauthorizedError } from "../errors/Unauthorized-error";
import UserAccountResource from "../resource/user-account-resource";
import { Cache } from "../internal/cache";

const USER_ACCOUNT_COLLECTION_NAME = "user-accounts";

export type User = {
  phone: string;
  firstName: string; // Automatically generated for guests, user defined for registered users.
  lastName: string; // Automatically generated for guests, user defined for registered users.
};

export type Account = {
  cardNumber: string;
  expiration: Date;
  pin: string;
  balance: number;
};

export type UserAccount = {
  _id?: ObjectId;
  user: User;
  account: Account;
};

const createUserPayload = zod.object({
  lastName: zod.string(),
  firstName: zod.string(),
  phone: zod.string().length(10),
});

const accountPayload = zod.object({
  pin: zod.string().length(4),
  cardNumber: zod.string().length(16),
  expiration: zod.date().min(new Date()),
  balance: zod.number().min(0),
});

const createUserAccountSchema = zod.object({
  user: createUserPayload,
  account: accountPayload,
});

export type CreateUserAccountPayload = zod.infer<
  typeof createUserAccountSchema
>;

const getUserCardInfoSchema = zod.object({
  cardNumber: zod.string().length(16),
  pin: zod.string().length(4),
});

export type GetUserCardInfoPayload = zod.infer<typeof getUserCardInfoSchema>;

export interface UserAccountControllerType {
  getUserCardInfo(
    payload: GetUserCardInfoPayload
  ): Promise<UserAccountResource | null>;
  createUserAccount(
    payload: CreateUserAccountPayload
  ): Promise<UserAccountResource>;
}

export class UserAccountController implements UserAccountControllerType {
  constructor(
    private readonly db: MongoClient,
    private readonly cache: Cache,
    private readonly logger: Logger
  ) {}

  public async getUserCardInfo(
    payload: GetUserCardInfoPayload
  ): Promise<UserAccountResource | null> {
    const { success, data, error } = getUserCardInfoSchema.safeParse(payload);
    if (!success) {
      this.logger.error("Invalid payload", data);
      throw new ValidationError("Payload validation failed", error);
    }

    const { cardNumber, pin } = data;

    try {
      const userAccount = await this.db
        .db(DATABASE_NAME)
        .collection<UserAccount>(USER_ACCOUNT_COLLECTION_NAME)
        .findOne({ cardNumber });

      if (!userAccount) {
        this.logger.error("User account not found");
        throw new NotFoundError("User account not found");
      }

      if (!verifyHash(pin, userAccount.account.pin, APP_SECRET!)) {
        this.logger.error("Invalid pin");
        throw new UnauthorizedError("Invalid pin");
      }

      userAccount.account.cardNumber = await decryptField(
        this.db,
        userAccount.account.cardNumber
      );

      return new UserAccountResource(userAccount);
    } catch (error) {
      this.logger.error(
        { "db.collection": USER_ACCOUNT_COLLECTION_NAME, error: error },
        "Failed to find user account by card number"
      );
      const err = handleDataLayerError(error as Error);
      throw err;
    }
  }

  public async createUserAccount(
    payload: CreateUserAccountPayload
  ): Promise<UserAccountResource> {
    const { success, data } = createUserAccountSchema.safeParse(payload);
    if (!success) {
      this.logger.error("Invalid payload", data);

      throw new ValidationError(
        "Payload validation failed",
        createUserAccountSchema
      );
    }

    const { user, account } = data;

    account.pin = hash(account.pin, APP_SECRET!);
    account.cardNumber = await encryptField(
      this.db,
      this.cache,
      account.cardNumber
    );
    account.expiration.setHours(23, 59, 59, 999); // Set to end of day

    const userAccount: UserAccount = {
      user,
      account,
    };

    try {
      const result = await this.db
        .db(DATABASE_NAME)
        .collection<UserAccount>(USER_ACCOUNT_COLLECTION_NAME)
        .insertOne(userAccount);

      const account = await this.db
        .db(DATABASE_NAME)
        .collection<UserAccount>(USER_ACCOUNT_COLLECTION_NAME)
        .findOne({ _id: result.insertedId });

      return new UserAccountResource(account!);
    } catch (error) {
      this.logger.error(
        { "db.collection": USER_ACCOUNT_COLLECTION_NAME, error: error },
        "Failed to create user account"
      );
      const err = handleDataLayerError(error as Error);
      throw err;
    }
  }
}
