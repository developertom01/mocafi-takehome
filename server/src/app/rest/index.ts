import { Router } from "express";
import { Logger } from "../../internal/logger";
import { Controller } from "../../controller";
import errorHandlerMiddleware from "./middleware/error-handler-middleware";
import registerUserRoutes from "./routes/user.route";
import registerAuthenticationRoutes from "./routes/authentication.route";
import { AsyncLocalStorage } from "async_hooks";

export function registerRestRoutes(
  controller: Controller,
  localStorage: AsyncLocalStorage<any>,
  logger: Logger
): Router {
  const apiRouter = Router();

  apiRouter.all("/", (_, res) => {
    res.json({ message: "Api server is running" });
  });

  apiRouter.use("/users", registerUserRoutes(controller, logger));
  apiRouter.use("/auth", registerAuthenticationRoutes(controller, logger));

  // Error handler
  apiRouter.use(errorHandlerMiddleware);

  return apiRouter;
}
