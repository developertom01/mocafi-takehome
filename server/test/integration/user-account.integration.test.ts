import request from "supertest";
import { AsyncLocalStorage } from "async_hooks";
import { LocalStorage } from "../../src/utils/types";
import {
  APP_SECRET,
  DATABASE_NAME,
  DATABASE_URL,
} from "../../src/config/app-config";
import MongoDbDatabase from "../../src/internal/database";
import { setupController } from "../../src/controller";
import { registerRestRoutes } from "../../src/app/rest";
import { ExpressRestServer } from "../../src/internal/rest-server";
import {
  GetUserCardInfoPayload,
  UserAccount,
} from "../../src/controller/user-accounts";
import { encryptField, hash } from "../../src/utils/cryptography";
import { getMockLogger } from "./utils";

describe("User Account Integration Test", () => {
  beforeEach(async () => {
    const [logger, logs] = getMockLogger();
    const localStorage = new AsyncLocalStorage<LocalStorage>();
    await MongoDbDatabase.instantiate(DATABASE_URL!);

    await MongoDbDatabase.connect();
    const controller = setupController(MongoDbDatabase.dbm, logger);

    const routes = registerRestRoutes(controller, logger);

    const restServer = new ExpressRestServer(routes, localStorage, logger);
    restServer.app.use((req, _, next) => {
      logger.info({ req });
      next();
    });

    expect.setState({
      app: restServer.app,
      database: MongoDbDatabase.dbm,
      logs,
    });
  });

  afterEach(async () => {
    // Delete all data in the database
    MongoDbDatabase.dbm?.db(DATABASE_NAME).dropDatabase();

    await MongoDbDatabase.disconnect();
  });

  test("/user-account/info", async () => {
    const { app, database, logs } = expect.getState();
    // Insert a user account
    const expiration = new Date();
    expiration.setFullYear(expiration.getFullYear() + 1);

    const account: UserAccount = {
      account: {
        balance: 10,
        cardNumber: await encryptField(database, "1234567890123456"),
        expiration,
        pin: hash("1234", APP_SECRET!),
      },
      user: {
        firstName: "John",
        lastName: "Doe",
        phone: "1234567890",
      },
    };

    await database
      .db(DATABASE_NAME)
      .collection("user-accounts")
      .insertOne(account);

    const payload: GetUserCardInfoPayload = {
      cardNumber: "1234567890123456",
      pin: "1234",
    };

    const expectedUserIfo = {
      phone: "1234567890",
      firstName: "John",
      lastName: "Doe",
    };
    const expectedAccountInfo = {
      cardNumber: "1234567890123456",
      expiration: `${expiration.getMonth() + 1}${expiration.getFullYear()}`,
      balance: 10,
    };

    const response = await request(app)
      .post("/api/user-accounts/info")
      .send(payload)
      .expect(200);

    expect(logs).toMatchSnapshot();
    expect(response.body).toMatchObject({
      user: expectedUserIfo,
      account: expectedAccountInfo,
    });
  });
});
