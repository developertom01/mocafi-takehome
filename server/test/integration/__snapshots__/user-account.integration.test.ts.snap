// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Account Integration Test /user-account/info 1`] = `[]`;
